api/METRICS.md

```diff
@@ -0,0 +1,91 @@
+# API proposal for metrics
+
+## Example
+
+```python
+# For most sklearn metrics, we would have their group version that returns a Bunch with fields
+#   * overall: overall metric value
+#   * by_group: a dictionary that maps sensitive feature values to metric values
+
+summary = accuracy_score_by_group(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+
+# Exporting into pd.Series or pd.DataFrame in not too complicated
+
+series = pd.Series({**summary.by_group, 'overall': summary.overall})
+df = pd.DataFrame({"model accuracy": {**summary.by_group, 'overall': summary.overall}})
+
+# Several types of scalar metrics for group fairness can be obtained from `summary` via transformation functions
+
+acc_difference = difference_from_summary(summary)
+acc_ratio = ratio_from_summary(summary)
+acc_group_min = group_min_from_summary(summary)
+
+# Most common disparity metrics should be predefined
+
+demo_parity_difference = demographic_parity_difference(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+demo_parity_ratio = demographic_parity_ratio(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+eq_odds_difference = equalized_odds_difference(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+
+# For predefined disparities based on sklearn metrics, we adopt a consistent naming conventions
+
+acc_difference = accuracy_score_difference(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+acc_ratio = accuracy_score_ratio(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+acc_group_min = accuracy_score_group_min(y_true, y_pred, sensitive_features=sf, **other_kwargs)
+```
+
+## Functions
+
+*Function signatures*
+
+```python
+metric_by_group(metric, y_true, y_pred, *, sensitive_features, **other_kwargs)
+# return the summary for the provided metrics
+
+make_metric_by_group(metric)
+# return a callable object <metric>_by_group:
+# <metric>_by_group(...) = metric_by_group(<metric>, ...)
+
+# Transformation functions returning scalars
+difference_from_summary(summary)

```

Comment:
How do all of these behave when presented with non-scalar metrics?

Reply:
I don't think we necessarily need to have a well-defined behavior for non-scalars. For `numpy.ndarray`, we could go with the pointwise operations.

Reply:
I have a severe allergy to undefined behaviour in programming languages. However, I'm not sure what would be considered the Pythonic thing to do.

Reply:
for start, let's just support scalars and throw exceptions otherwise.

Reply:
Suits me